### ISSUE TITLE
Getting ratelimited

### ISSUE THREAD

**Issue by Koyamie** [discordjs/discord.js#3560]
**Please describe the problem you are having in as much detail as possible:**
I am getting ratelimited by discord for an unknown reason
As said here it should never happen but I do not know why this is happening for me
https://github.com/discordjs/discord.js/blob/6be5051f9296fe9a10549f2f9a11c18776b1b617/src/rest/RequestHandler.js#L149

For more information, it's happening frequently at 00:00 GMT

I am also having a lot of 500 errors when this is happening.

**Include a reproducible code sample here, if possible:**
```js
429 hit on route /gateway/bot

Code : 500
AbortError: The user aborted a request.
at RequestHandler.execute (/home/koyamie/Koya/node_modules/discord.js/src/rest/RequestHandler.js:106:9)
at runMicrotasks (<anonymous>)
at processTicksAndRejections (internal/process/task_queues.js:93:5)
```

**Further details:**
- discord.js version: lastest master
- Node.js version: 12.13.0
- Operating system: debian 9
- Priority this issue should have – please be realistic and elaborate if possible:

- [x] I have also tested the issue on latest master, commit hash:

**Comment by Koyamie** [comment]
Hello, kinda high issue, happening everyday since I updated the repo (was on a september commit)

**Comment by Fyko** [comment]
Is it possible you hit the max of 1000 logins per day? Have you tried listening on the debug event?

**Comment by Koyamie** [comment]
> Is it possible you hit the max of 1000 logins per day? Have you tried listening on the debug event?

Nope, when I am hitting 1k logins a day its different, I can't login for 24 hours and debugging tells how many time I have to wait, here I am getting a 429 "Too many requests" and debugging only shows this :
"429 hit on route /gateway/bot"

**Comment by Monbrey** [comment]
If you need help with discord.js installation or usage, please go to the discord.js Discord server instead: https://discord.gg/bRCvFy9
This issue tracker is only for bug reports and enhancement suggestions.

**Comment by Koyamie** [comment]
> If you need help with discord.js installation or usage, please go to the discord.js Discord server instead: https://discord.gg/bRCvFy9
> This issue tracker is only for bug reports and enhancement suggestions.
> 
> Since this is happening at exactly 0:00 it would seem there's some scheduled task running at this time causing it. We can investigate further on Discord.

Yeah but, as the line is saying "This should never happen" I guess this is related to the lib.

**Comment by DevYukine** [comment]
i could imagine that this might be related to multiple Shards/Clients fetching the /bot/gateway endpoint at the same time and since they dont know of each other they do too many requests and hit the ratelimit, are you sure you dont fetch the /bot/gateway endpoint on your own? because d.js should only ever fetch that on startup and i can't imagine that your bot restarts everytime on 0:00 GMT

**Comment by Koyamie** [comment]
> i could imagine that this might be related to multiple Shards/Clients fetching the /bot/gateway endpoint at the same time and since they dont know of each other they do too many requests and hit the ratelimit, are you sure you dont fetch the /bot/gateway endpoint on your own? because d.js should only ever fetch that on startup and i can't imagine that your bot restarts everytime on 0:00 GMT

I think too but I have no reason fetching this endpoint on my own.
Yup my bot isn't restarting everytime but at 0:00 GMT my game timestamps reset (like the tatsumaki daily and reputation) but there is no reason that it fetch this endpoint 🤔
This is weird

**Comment by DevYukine** [comment]
Okay i just looked in the src and saw that d.js actuall does call the /bot/gateway endpoint everytime it reconnects (this shouldn't be like that) and since you mentioned that you get these every day at 0:00 GMT this probaly means that your bot reconnects every day at that time (maybe related to your host?) sadly d.js does always fetch the session endpoint even when resuming a session but vlad actuall fixed that in #3393 so you could give this pr a try since it should fix your issue (he changed it so it only fetches /gateway/bot if no session_id id present).

Also you should probaly check why the bot reconnects everytime at 00:00 GMT, do you maybe change your ip or does your isp/host do that? i would recommend looking into that.

Even tho #3393 probaly fixes this issue i will open another one because the current implementation is very poor designed for bots using any ShardingManager and this needs to be changed since we currently do not handle ratelimits for /gateway/bot across shards.

**Comment by Koyamie** [comment]
> Okay i just looked in the src and saw that d.js actuall does call the /bot/gateway endpoint everytime it reconnects (this shouldn't be like that) and since you mentioned that you get these every day at 0:00 GMT this probaly means that your bot reconnects every day at that time (maybe related to your host?) sadly d.js does always fetch the session endpoint even when resuming a session but vlad actuall fixed that in #3393 so you could give this pr a try since it should fix your issue (he changed it so it only fetches /gateway/bot if no session_id id present).
> 
> Also you should probaly check why the bot reconnects everytime at 00:00 GMT, do you maybe change your ip or does your isp/host do that? i would recommend looking into that.
> 
> Even tho #3393 probaly fixes this issue i will open another one because the current implementation is very poor designed for bots using any ShardingManager and this needs to be changed since we currently do not handle ratelimits for /gateway/bot across shards.

This is was I thought, I will give a try to this PR and check tonight if it is happening again and will update the issue. My bot is not reconnecting at 0:00 this is why I do not understand 😯

**Comment by DevYukine** [comment]
> > Okay i just looked in the src and saw that d.js actuall does call the /bot/gateway endpoint everytime it reconnects (this shouldn't be like that) and since you mentioned that you get these every day at 0:00 GMT this probaly means that your bot reconnects every day at that time (maybe related to your host?) sadly d.js does always fetch the session endpoint even when resuming a session but vlad actuall fixed that in #3393 so you could give this pr a try since it should fix your issue (he changed it so it only fetches /gateway/bot if no session_id id present).
> > Also you should probaly check why the bot reconnects everytime at 00:00 GMT, do you maybe change your ip or does your isp/host do that? i would recommend looking into that.
> > Even tho #3393 probaly fixes this issue i will open another one because the current implementation is very poor designed for bots using any ShardingManager and this needs to be changed since we currently do not handle ratelimits for /gateway/bot across shards.
> 
> This is was I thought, I will give a try to this PR and check tonight if it is happening again and will update the issue. My bot is not reconnecting at 0:00 this is why I do not understand 😯

Are you sure about this? because it very much looks like it (reconnecting does not only include re-identifying but also resuming a session) can you make sure and listen to the `shardReconnecting` event (e.g log it to the console or somehow notify yourself) and update this post once you tested it?

**Comment by Koyamie** [comment]
![D2990BDE-ED90-4914-A960-DA06B9A8CB56](https://user-images.githubusercontent.com/19957779/68086060-f3ec6600-fe47-11e9-8080-955f8a1969a3.png)

I am logging this event in this channel, this is tonight's events, none of the shards reconnected at 1am CET (0:00 GMT) so yeah this is weird.

**Comment by Koyamie** [comment]
Update : I started debugging tonight to see what happen at 0:00 GMT, I found what in my bot was getting me ratelimited for 1h.

I was hitting 429 very very many times but with different enpoints, the most I saw was `/channels/ID/messages` this one should be fixed in my code but still hitting 429 even if it should never happen.

I havn't tried the PR because I got no time yesterday.

Anyway, I'll keep this issue open until we get a fix on master.

**Comment by iCrawl** [comment]
You are being very vague about this.

`/channels/Id/messages`, doesn't mean a lot. This needs more information on your part, maybe even some code samples of what your usage is and why you are hitting it.

It would be nice if this is getting a bit more context.

**Comment by DevYukine** [comment]
If you get ratelimited on other routes e.g `/channels/ID/messages` you might hit the Global Ratelimit and then hitting a 429 is normal but then there is still the question why you hit `/gateway/bot` aswell at that time 🤔

**Comment by Koyamie** [comment]
> You are being very vague about this.
> 
> `/channels/Id/messages`, doesn't mean a lot. This needs more information on your part, maybe even some code samples of what your usage is and why you are hitting it.
> 
> It would be nice if this is getting a bit more context.

Yes very sorry will explain better as soon as I can



> If you get ratelimited on other routes e.g `/channels/ID/messages` you might hit the Global Ratelimit and then hitting a 429 is normal but then there is still the question why you hit `/gateway/bot` aswell at that time 🤔

Maybe hitting it too much result to a gateway ban for 1 hour (Can't connect for 1 hour everytime it happens)

**Comment by Koyamie** [comment]
Hello, this is happening today but in different way, restarted my bot after making changes and now bot is having very difficulties to connect and I am hitting 429 way too much (I should never hit 429 remember)
Any help ?
![2019-12-01_1832](https://user-images.githubusercontent.com/19957779/69919847-99afe680-1481-11ea-982b-7bd55f6734db.png)

**Comment by Koyamie** [comment]
Hello, it happened again tonight with discord increased API response time. Any news for this ?

**Comment by Koyamie** [comment]
UP

**Comment by Nooc34** [comment]
Having the same issue here now. 
`429 hit on route /gateway/bot`

Have 0 idea this has never happened before.

very frustrating

**Comment by vladfrangu** [comment]
If it happens once or a few times and you're sharded across multiple processes, it happens. If it's spammed, it's an issue.

d.js version, node.js version please!

**Comment by Nooc34** [comment]
Discord.js ^12.2.0
Node.js 11.x

It isnt spammed, just happens every minute or so.

**Comment by Monbrey** [comment]
> Discord.js ^12.2.0
> Node.js 11.x

Discord.js v12 has a minimum Node version of v12 also. This doesn't line up

**Comment by Nooc34** [comment]
it worked for 3 months originally?

**Comment by SudhanPlayz** [comment]
Even me bot the problem. But i was just connected like 10 times

```
429 hit on route /gateway/bot
429 hit on route /gateway/bot
429 hit on route /gateway/bot
429 hit on route /gateway/bot
429 hit on route /gateway/bot
429 hit on route /gateway/bot
429 hit on route /gateway/bot
429 hit on route /gateway/bot
429 hit on route /gateway/bot
429 hit on route /gateway/bot
429 hit on route /gateway/bot
429 hit on route /gateway/bot
429 hit on route /gateway/bot
429 hit on route /gateway/bot
429 hit on route /gateway/bot
429 hit on route /gateway/bot
429 hit on route /gateway/bot
429 hit on route /gateway/bot
429 hit on route /gateway/bot
```

**Comment by SudhanPlayz** [comment]
My project is not working on Any Bot Tokens Pls help everybots getting strucked at 
```js
Provided token: NzEwODE2MTI3MTA5OTU1NjA1.XsOUQw.***************************
Preparing to connect to the gateway...
429 hit on route /gateway/bot
```
I Don't what I want to do. also I just logging one time in my project

**Comment by Koyamie** [comment]
> My project is not working on Any Bot Tokens Pls help everybots getting strucked at
> 
> ```js
> Provided token: NzEwODE2MTI3MTA5OTU1NjA1.XsOUQw.***************************
> Preparing to connect to the gateway...
> 429 hit on route /gateway/bot
> ```
> 
> I Don't what I want to do. also I just logging one time in my project

You are IP banned for 1 hour, just wait and retry.

**Comment by SudhanPlayz** [comment]
Thanks! I will try after 1 hour

**Comment by SudhanPlayz** [comment]
The ip was banned forever i think even it not worked now. I was using glitch.com to run my bot

**Comment by CatsMiaow** [comment]
- Node.js v12.18.2
- discord.js v12.3.1

Suddenly the bot did not respond without any error, and when I restarted the bot and tried to client login again, it did not login without any error.
When I called the Discord API directly from the server, I found out that Discord blocked it.

`curl https://discord.com/api/v6`
```json
{"code": 0, "message": "You are being blocked from accessing our API temporarily due to exceeding our rate limits frequently. Please read our docs at https://discord.com/developers/docs/topics/rate-limits to prevent this moving forward."}
```
When I tried again after an hour, it would be nice if the bot was logged in but got an error saying it was blocked in Discord.

**Comment by Fyko** [comment]
I was running into this problem for... a very long time (dating back to mid march). Here are some [sparse error](https://paste.nomsy.net/urefofipih.md) [logs](https://paste.nomsy.net/xifoqetidu.md). I reached out to a Discord staff member in the Discord Developers server and they said a bot on my network was consistently hitting rate limits and that led to a one hour CloudFlare ban. 

In the end, I decided to migrate two of my larger bots (300 & 900 guilds) from my local machines ~ into the cloud ~ and that seemed to mitigate the problem entirely -- haven't ran into any bans since. 

tldr: hoarding multiple bots on a single network should be avoided

**Comment by DjDeveloperr** [comment]
For me, it keeps spamming. It's a serious problem for me, bot goes down for an hour.
Seems to be a bug on Discord's end, since the retry after header is incorrect and so the lib keeps trying to reconnect.
This starts happening with a similar rate limit, on channels something route, it spams it and then gets IP banned. I'm sure discord.js is handling rate-limits correctly and I'm not doing any API Request outside lib. 
In that channel rate limit too, even after the retry after time span ends, it's still getting rate limit headers, so probably Discord giving out wrong values for rate-limits? 
This never happened to me before, just started happening since a week and it's like happening every 2-3rd day.
You might be curious how am I so sure about "incorrect rate limit headers", I had modified src/RequestHandler.js and added some more info to the log from the headers (and body too, showing that its _IP Banned_).
If you're curious what code I added, here it is:
```js
    } else if (res.status === 429) {
      // A ratelimit was hit - this should never happen
      this.queue.unshift(item);
      let body = await res.text();
      this.manager.client.emit('debug', `429 hit on route ${item.request.route} - rate limit ${this.limit} - retry after ${this.retryAfter} - reset ${this.reset} - remainaing ${this.remaining} - global ${(this.manager.globalTimeout}\nResponse Body: ${body}`);
      await Util.delayFor(this.retryAfter);   
      return this.run();
```
About the logs I see in console from this rate limit:
![image](https://user-images.githubusercontent.com/43033058/91970446-4c74e400-ed35-11ea-8fa1-10d49cd39abd.png)

Hope it helps to identify.. would appreciate if this issue is resolved..

**Comment by Koyamie** [comment]
Happened to me today again too.
Usually it is mostly due to the sharding not having the same headers because of the large amounts of shards and the speed of the requests.

But today my logs are showing A SUPER HUGE spam of ` [RATELIMIT] GET /gateway/bot` requests. Probably shards reconnecting but this went OVER 10k 429s in under 10 minutes which is very high.

**Comment by DjDeveloperr** [comment]
This bug is annoying, my bot just got rate limited again. I opened logs on time, and found that it was spamming some rate limit on messages. But due to the headers being incorrect, it had kept retrying and got IP banned at last, for an hour. I think this is a bug related to Discord API, no?

**Comment by msciotti** [comment]
Sorry to sort of butt in here—for the folks having this error, are you running your bot on multiple processes on the same machine? I'm unsure if DJS (or other libraries) support coordinating cross-process rate limits out of the box/without some config options enabled.

**Comment by DjDeveloperr** [comment]
Nope, my bot is on single process and not sharded. 
It happened again (some minutes ago)...

**Comment by kyranet** [comment]
Note: it would be helpful to also include the version of discord.js you are using, e.g. v12.3.1 - this way we can check whether you are using old versions (specially v11) or latest one.

Also, you can still reproduce this by restarting your bot quickly after it logged in.

> Sorry to sort of butt in here—for the folks having this error, are you running your bot on multiple processes on the same machine? I'm unsure if DJS (or other libraries) support coordinating cross-process rate limits out of the box/without some config options enabled.

discord.js doesn't have persistent nor coordinates cross-process rate limits, such thing would require an external service (i.e. Redis) to keep the state and help synchronizing all clients.

**Comment by DjDeveloperr** [comment]
I'd appreciate if this issue is given a priority. It's very disturbing \:v
> Note: it would be helpful to also include the version of discord.js you are using, e.g. v12.3.1 - this way we can check whether you are using old versions (specially v11) or latest one.
I had mentioned before, but, again mentioning  it,
Discord.js Version => 12.3.1
Node.js Version => 12.18.2

Also, I did a hacky work-around which at least doesn't spam after getting IP Banned. I noticed the RetryAfter header is in seconds when this ban is hit, I just modified it a bit again.
In src/RequestHandler.js
```js
    } else if (res.status === 429) {
      // A ratelimit was hit - this should never happen
      this.queue.unshift(item);
      let body = await res.text();
      this.manager.client.emit('debug', `429 hit on route ${item.request.route} - rate limit ${this.limit} - retry after ${this.retryAfter} - reset ${this.reset} - remainaing ${this.remaining} - gt ${(this.manager||{globalTimeout:0}).globalTimeout}\nResponse Body: ${body}`);
      await Util.delayFor(this.retryAfter * (body.includes("You are being blocked") ? 1000 : 1));
      return this.run();
```
It would wait until the time is over correctly and not spam 429s.
Still not able to find out the cause of this.

**Comment by vladfrangu** [comment]
> I just noticed that the RetryAfter header is in seconds when this rate-limit is hit! I got to know by observing the change in RetryAfter header, because the lib is treating it as given in milliseconds, while its coming in seconds from the API.

Unless you changed the REST api version to v8 yourself via client options, Discord returns the header as ms instead of seconds right now on the API version we support and use (v7).. 

Now, I wonder if this is a Cloudflare ban instead of a Discord ban, @msciotti if you could enlighten us that'd be awesome (based on the JSON body response and the retry-after header being seconds, my gut says it is)! If this IS a Cloudflare ban, can we get a code that isn't `0` in the JSON body, if possible? 😅 Something like the current global ratelimit response would be amazing (or at least a different code so we can check by it and globally limit the entire library from doing further requests)

For @DjDeveloperr, since I noticed you modified the code, can you on 429's log the `via` header from `res.headers`? Discord replies will always have that header (usually with the content `1.1 google`; Cloudflare bans don't have that).. We *currently* don't handle the difference between Clouflare and Discord bans, which may cause this issue

**Comment by DjDeveloperr** [comment]
I'm pretty late here - I had found the cause lol. It was because of webhooks, which sent "logs" so fast and their rate limits were not handled. 
> For @DjDeveloperr, since I noticed you modified the code, can you on 429's log the via header from res.headers? Discord replies will always have that header (usually with the content 1.1 google; Cloudflare bans don't have that).. We currently don't handle the difference between Clouflare and Discord bans, which may cause this issue
Those variables are already from headers.

> Unless you changed the REST api version to v8 yourself via client options, Discord returns the header as ms instead of seconds right now on the API version we support and use (v7)..
I didn't change, but I had to multiply the retryAfter with 1000 for avoiding the additional spam. :/

**Comment by vladfrangu** [comment]
> I didn't change, but I had to multiply the retryAfter with 1000 for avoiding the additional spam. :/
That's most likely because the API ban comes from Cloudflare (or so I assume)... Hence why the multiply by 1000 worked; this is why I asked if you can log the `Via` header 😃

**Comment by Koyamie** [comment]
Yes, the API ban is coming from Cloudflare, we also call it a CF ban.

For my case, I will explain how it was happening:

My bot was hitting the 50req/s limit set by Discord, this was causing a Global Ratelimit. The shard that was receiving the ratelimit was correclty handling it but other ones could not, this is how I was getting massive 429's (with 320 shards means a lot) because they were still sending request while being ratelimited.

When this was happening it could solve itself after few seconds and was hitting around 300-400 429's but sometimes it was reaching the 10k limit in few seconds.

The last few weeks, it was happening 1 time a day and sometimes 2.

**Comment by DjDeveloperr** [comment]
> I asked if you can log the `Via` header 😃

Sorry, but since I removed the webhook logs Rate limit didn't get hit since then, I assume its fixed now. It was a problem on my end.

**Comment by Koyamie** [comment]
I can confirm that using webhooks to logs creates high rate of 429's.

Im reaching around 100 ratelimits in past 10 minutes when I enable them.

**Comment by Deivu** [comment]
I got banned 2x this day thanks to this 429 at `gateway/bot`. I disabled webhook logs, still getting cloudflare banned.
To solve this, I tried the code snippet by @DjDeveloperr like so
```js
const body = await res.json();
this.manager.client.emit('debug', `429 hit on route ${request.route} | ${this.retryAfter} idk if this ms or seconds`);
await Util.delayFor(this.retryAfter * (body?.message?.includes("You are being blocked") ? 1000 : 1));

```
I've been down for a bit of time, around 4 hours already due to this issue. This issue could potentially prevent big bots from logging in 👀

**Comment by Wilco2998** [comment]
I also got a rate limit and only discovered it via the debug eventemitter, it would be nice if the "ratelimit" event would really emit when 429 was given as a response

**Comment by atomheartother** [comment]
My bot is down for hours at a time because of this issue, I'm running 7 shards and getting rate-limited almost immediately on boot. I don't understand why it suddenly started getting rate-limited a week ago, and again today, is this really on Discord's end?

Also very unhelpful, ShardingManager appears to just throw an unhandledexception when this occurs, it does not display any sort of 429 error, which made this extremely confusing to debug...

**Comment by Poly-Pixel** [comment]
> My bot is down for hours at a time because of this issue, I'm running 7 shards and getting rate-limited almost immediately on boot. I don't understand why it suddenly started getting rate-limited a week ago, and again today, is this really on Discord's end?
> 
> Also very unhelpful, ShardingManager appears to just throw an unhandledexception when this occurs, it does not display any sort of 429 error, which made this extremely confusing to debug...

Is this still an issue? If so, what discord.js/node.js version?

**Comment by muhimur9049** [comment]
> Is this still an issue? If so, what discord.js/node.js version?

Recently just got the 429 error. I have no idea what caused it. It could be because of the webhooks for errors my bot uses, or maybe it was using the `.fetch` method too often when fetching a channel (?).

I'm going to try and restart the bot in about an hour to see if it goes away or not by then.

**Comment by Jiralite** [comment]
Closing this as it is a stale issue. Feel free to make a new issue if this problem still persists.
